### ISSUE TITLE
The exp claim should be optional

### ISSUE THREAD

**Issue by cloutiertyler** [Keats/jsonwebtoken#190]
According to the RFC, the `exp` claim should be optional: https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.4

> 4.1.4.  "exp" (Expiration Time) Claim
> ... Use of this claim is OPTIONAL.

I saw that you suggested that the user just set an expiration very far in the future. This works for new tokens, however it doesn't work for tokens created previously (e.g. if you're migrating tokens to a new service).

**Comment by Keats** [comment]
We could make it optional but is there anyone using JWTs without exp?

**Comment by cloutiertyler** [comment]
We were issuing some tokens without `exp` on our old service because they were stored securely and we can also manually expire them, but regardless of that I think the library should implement the RFC to spec. In my experience, even small deviations from the RFC can have unintended consequences that aren't immediately obvious (migrating from an old service is just one example).

Fortunately in this case the RFC isn't ambiguous at all and since it's going from required to optional it's a non-breaking change for the library. I can definitely understand the desire for wanting `exp` to be mandatory, since it's good practice, but I don't think it's worth potentially preventing someone from using the library because their systems rely on it's optionality.

Maybe there's a way for the library to encourage its use, without absolutely requiring it.

**Comment by Keats** [comment]
> Maybe there's a way for the library to encourage its use, without absolutely requiring it.

**Comment by Keats** [comment]
Actually looking at it again, if you set `Validation.validate_exp` to `false` it should be working fine?

**Comment by brocaar** [comment]
@Keats I think I'm running into the same issue as @cloutiertyler. I don't think the `validate_exp = false` is the solution. 

In my use-case, two different types of claims can be issued, claim A will have an `exp` set (and must be validated), claim B does not have an `exp` set (and there is no expiration to validate).

Currently this is not possible as `validate_exp = true` will invalidate claim B (as no `exp` is considered an expired claim if I'm not mistaken), `validate_exp = false` is not an option as claim A with an `exp` must be validated. You only know which claim was received after validating it, so you can't adapt your validation settings per claim type.

My thoughts:

Is there an use-case for the `validate_exp` option? In other words, if you have a claim with an `exp` which has already expired, is there an use-case to let it pass the validation?

By removing the `validate_exp` option and only validating the `exp` if it is set, users still have the flexibility to use it or not, based on if they include the `exp` into the generated claims or not (which makes it optional to use). I think the same applies to `validate_nbf`.

An alternative would be to keep these options (enabled by default), so that for debugging purposes the validation of `exp` and `nbf` can be skipped. In this case the `exp` (and `nbf`) would only be validated if these fields are set in the claim.

**Comment by Keats** [comment]
> Is there an use-case for the validate_exp option?

To ensure people are actually using an `exp` claims. I've seen many people get tripped up by `validate_exp` because they didn't realise they had to set an exp claim themselves.
Also in pyjwt they have only validate the claims if they are present but that means there is another validation called `require` to require specific claims to be present. For a common usecase, that makes the validation setup less clear imo. I think the vast majority of JWTs will know whether they have an `exp` or not and odd cases like yours where you don't know, you should probably handle it yourselves based on other claims. Otherwise your claim A could stop including `exp` and you would never know.

**Comment by brocaar** [comment]
> I've seen many people get tripped up by validate_exp because they didn't realise they had to set an exp claim themselves

Maybe this could be better facilitated at the claim generating side than the validating side? According to the JWT specs, they are still generating valid tokens, considered valid by pyjwt, jwt-go and others but invalid by jsonwebtoken. One option could be to provide an extendible standard claims struct which users may use (still the option to use their own custom claims), which automatically comes with the common claims and sets the `exp` to a user-provided TTL? 

I agree with you that the the library should be provided with sane defaults that provide generating secure tokens and secure validation, but I think my use-case is not uncommon as the `exp` (and `nbf`) fields are specified as optional in the specification and therefore they might be present or not.

> Also in pyjwt they have only validate the claims if they are present but that means there is another validation called require to require specific claims to be present

I don't think it would be a bad approach to be honest. You could drop the `validate_exp` and `validate_nbf` options (validate `exp` and `nbf` only if present) and have one option to configure the mandatory claims. This could still be a sane default that promotes the usage of the `exp` and `nbf` claims (or only `exp`). For a "normal" use-case it would mean there is nothing to customize. In my case it would mean I would remove `exp` (and `nbf`) from this list of mandatory claims, which would make them optional (only validated when present).


Please note that the jwt-go library has the same implementation, these claims are only validated when present:
https://github.com/dgrijalva/jwt-go/blob/master/claims.go#L30

```go

// Validates time based claims "exp, iat, nbf".
// There is no accounting for clock skew.
// As well, if any of the above claims are not in the token, it will still
// be considered a valid claim.
func (c StandardClaims) Valid() error {
	vErr := new(ValidationError)
	now := TimeFunc().Unix()

	// The claims below are optional, by default, so if they are set to the
	// default value in Go, let's not fail the verification for them.
	if c.VerifyExpiresAt(now, false) == false {
		delta := time.Unix(now, 0).Sub(time.Unix(c.ExpiresAt, 0))
		vErr.Inner = fmt.Errorf("token is expired by %v", delta)
		vErr.Errors |= ValidationErrorExpired
	}

	if c.VerifyIssuedAt(now, false) == false {
		vErr.Inner = fmt.Errorf("Token used before issued")
		vErr.Errors |= ValidationErrorIssuedAt
	}

	if c.VerifyNotBefore(now, false) == false {
		vErr.Inner = fmt.Errorf("token is not valid yet")
		vErr.Errors |= ValidationErrorNotValidYet
	}

	if vErr.valid() {
		return nil
	}

	return vErr
}
```

**Comment by cloutiertyler** [comment]
> Actually looking at it again, if you set `Validation.validate_exp` to `false` it should be working fine?

This is not fine, unfortunately. I *do* want to validate tokens which have an `exp`, but I also need tokens that don't have an `exp` to not be validated. I need this on a per token basis. Not all validated or all not validated.

**Comment by cloutiertyler** [comment]
> That's mostly what I was trying to figure out. I'll see what i can do for v8.

Thank you!

> I try to not deviate too much unless it's something stupid in the spec (eg the `none` algorithm which is not supported in this library).

Agreed!

**Comment by Keats** [comment]
> This is not fine, unfortunately. I do want to validate tokens which have an exp, but I also need tokens that don't have an exp to not be validated. I need this on a per token basis. Not all validated or all not validated.

I'm a bit confused, does those tokens end up using the same code path? I always have different calls with pyjwt depending on what I am doing because I usually need to check different scopes etc and I do want to know if for some reason my oauth access token doesn't have `exp`, not silently validate it.

**Comment by brocaar** [comment]
In my case it would until and including the validation. Then after validation, based on the token claims the code path would split. I think this is a valid use-case.

> I do want to know if for some reason my oauth access token doesn't have exp, not silently validate it.

I believe that according to the specs, such oauth access token without `exp` is considered valid. To reject such tokens is the decision of your application logic (and I don't disagree it is a bad one btw). But technically such token is valid (assuming it passes the other tests).

I think the configurable list of mandatory claims would be a best solution to deal with all use-cases and be complaint with the JWT specs. Personally I don't mind if the `exp` is included by default or not, as long as this is configurable :)

**Comment by Keats** [comment]
> I believe that according to the specs, such oauth access token without exp is considered valid

**Comment by cloutiertyler** [comment]
> I do want to know if for some reason my oauth access token doesn't have exp, not silently validate it.

A JWT without an `exp` is valid and should be validated according to the spec. If this library doesn't do that, then it's a "KeatsJsonWebToken" library, and not a JsonWebToken library. That's obviously your choice, but please be aware this is not a correct implementation. That's more or less what I want to point out.

**Comment by Keats** [comment]
Oops the rest of my message didn't get sent :(

>A JWT without an exp is a valid and non-expired token and should successfully validate according to the spec.

It does currently, it just doesn't handle having using the same validation rules for tokens that have `exp` or not.
Oauth specs requires `exp` (as well as sub, iss, aud): https://datatracker.ietf.org/doc/html/rc7523#section-3 for the access tokens.

The reason I'm hesitant changing that is, continuing the oauth usecase, it will validate just fine invalid tokens (eg missing one of the required fields above) despite having `validate_exp` (or sub etc) set to `true`. So this requires setting the additional list of `mandatory` claims which is imo a bit confusing but I don't see another way. I guess it would need documentation to explain it well.

**Comment by brocaar** [comment]
> So this requires setting the additional list of mandatory claims

In my opinion a `mandatory_claims` list would replace the `validate_exp` and `validate_nbf` options. That would make it a lot easier to understand and to document. How I think this could work:

1. First validate that all items in `mandatory_claims` are available in the token. Note that `mandatory_claims` could contain any claim, also non-standardized application-specific ones (e.g. it is just a list of names).
2. Token contains `exp` claim? If yes, validate it has not expired.
3. Token contains `nbf` claim? If yes, make sure that `nbf` is in the past.
4. ...

As the library would only validate `exp` and `nbf` if present, there is no reason to set a `validate_exp` or `validate_nbf`. If the `exp` or `nbf` claims are present, you always want to validate these.

I don't mind if the `mandatory_claims` would by default include `exp` and `nbf`, as long this is configurable :) By including them, you could still offer secure defaults so that a missing `exp` would not go unnoticed without the reconfiguration of this `mandatory_claims` option :)

**Comment by Keats** [comment]
> In my opinion a mandatory_claims list would replace the validate_exp and validate_nbf options

That would break the symmetry with the other validations which I'm not too keen on. Plus you might want to require some fields but not have the lib handle the validation (I've seen some exotic JWT with local time instead of UTC for `exp`...). I think I'll just go with the required claims field + only validate if present.

**Comment by DanielJoyce** [comment]
I'm all for more flexible validation. Being able to use a function would be great too. I'd like to be able to validate azp (Authorized Party) as well.

Some factory methods on Validator for common scenarios would be useful too.

If we had validation functions, then @brocaar 's use case would be satisified as well.

**Comment by Keats** [comment]
This library will only handle things from the spec. If you have other claims you want to validate like the openid ones, it will have to be done by hand in your code.

**Comment by Keats** [comment]
Ok, a bit late but can people have a look at https://github.com/Keats/jsonwebtoken/pull/225 ?

**Comment by brocaar** [comment]
Hi @Keats, I think this hasn't been solved by #225.

A test-case that can be added to `validation.rs` that catches this issue would be:

```rust
    #[test]
    fn optional_exp() {
        let claims = json!({});
        let mut validation = Validation::new(Algorithm::HS256);
        validation.required_spec_claims = HashSet::new();
        validation.validate_exp = true;
        let res = validate(deserialize_claims(&claims), &validation);
        assert!(res.is_ok());
    }
```

`res` is `Err(Error(ExpiredSignature))`, but should be OK as `exp` is not required and should only be validated if present.

**Comment by Keats** [comment]
I think https://docs.rs/jsonwebtoken/latest/jsonwebtoken/struct.Validation.html#structfield.required_spec_claims is clear in terms of docs? You are asking it to validate the exp explicitly. `required_spec_claims` just checks for the presence of fields

**Comment by brocaar** [comment]
Then could you explain again how to make the `exp` optional, so that:

* If `exp` is present in the claim, validate that it didn't expire yet
* If no `exp` is present in the claim, do not throw an error

My understanding of #225 was that the `required_spec_claims` would handle the claims which must be present (which is why I set this to an empty set) and that setting `validate_exp` would validate the `exp` if present:

![image](https://user-images.githubusercontent.com/165497/153160925-7f52d141-7772-460e-95bf-c7e38db9097d.png)

**Comment by Keats** [comment]
ah my bad. Is the link above to docs.rs enough explanation?

**Comment by brocaar** [comment]
@Keats, not really. Could you please explain me how #225 solves this GitHub issue? As I explained earlier in this issue, I have tokens that have the `exp` set and tokens that don't have the `exp` set. Implementing two different validators is not an option as only after validation I know the purpose of these tokens. Could you explain what config to use when the `exp` must **only be validated if present?**

I do understand that `required_spec_claims` should not contain the `exp` claim as in this case it must always be present. However, I would in that case expect the `validate_exp` to only validate the expiration if the claim is present. I thought that was the purpose of #225 to allow the following scenarios:

* Optional `exp` (`exp` is not mandatory / validate `exp` if present)
  * `required_spec_claims = []`
  * `validate_exp = true`

* Required `exp` (`exp` is mandatory and validate `exp`)
  * `required_spec_claims = ["exp"]`
  * `validate_exp = true`

* Required `exp` (`exp` is mandatory but not validated)
  * `required_spec_claims = ["exp"]`
  * `validate_exp = false`

* Ignore `exp` completely (`exp` is not mandatory and is not validated if present)
  * `required_spec_claims = []`
  * `validate_exp = false`

**Comment by Keats** [comment]
Ah then the implementation doesn't do that then, I misunderstood.

**Comment by cloutiertyler** [comment]
I'm on the same page as @brocaar. I have the following tokens mixed together:

- tokens that don't expire
- tokens that do expire

Both can be valid, but the tokens that do expire are only valid if they're not expired. I basically want tokens that don't expire to act as if they have an infinite expiry time. I also cannot change any tokens that have already been created for backwards compatibility.

**Comment by Keats** [comment]
Honestly it's extremely unlikely that this crate will support _maybe_ verifying exp. That sounds like a recipe for disasters, if your tokens are being generated without `exp` they are still valid? Is there any JWT library in any language that works like that?

**Comment by brocaar** [comment]
@Keats, but you did propose to implement it this way ;-) (https://github.com/Keats/jsonwebtoken/issues/190#issuecomment-871674415)

> I think I'll just go with the required claims field + **only validate if present**.

> That sounds like a recipe for disasters

I disagree. If `validate_exp = true` (which is the default) would **only** validate if the `exp` claim is present **and** it would be a required claim by setting `required_spec_claims = ["exp"]` (which is also the default), then the default behavior does not change, as:

1. The presence of `exp` is mandatory
2. If `exp` is present (which is already enforced by 1.), it will be vaildated

> if your tokens are being generated without exp they are still valid?

Yes, they are.

> Is there any JWT library in any language that works like that?

The Go implementation does work like this. See: https://github.com/golang-jwt/jwt/blob/78a18c0808520e53b59e7a4396915430d51749f1/claims.go#L49


```go
// Valid validates time based claims "exp, iat, nbf".
// There is no accounting for clock skew.
// As well, if any of the above claims are not in the token, it will still
// be considered a valid claim.
func (c RegisteredClaims) Valid() error {
	vErr := new(ValidationError)
	now := TimeFunc()

	// The claims below are optional, by default, so if they are set to the
	// default value in Go, let's not fail the verification for them.
	if !c.VerifyExpiresAt(now, false) {
		delta := now.Sub(c.ExpiresAt.Time)
		vErr.Inner = fmt.Errorf("%s by %v", delta, ErrTokenExpired)
		vErr.Errors |= ValidationErrorExpired
	}

	if !c.VerifyIssuedAt(now, false) {
		vErr.Inner = ErrTokenUsedBeforeIssued
		vErr.Errors |= ValidationErrorIssuedAt
	}

	if !c.VerifyNotBefore(now, false) {
		vErr.Inner = ErrTokenNotValidYet
		vErr.Errors |= ValidationErrorNotValidYet
	}

	if vErr.valid() {
		return nil
	}

	return vErr
}
```

The same for Python:

https://pyjwt.readthedocs.io/en/stable/api.html?highlight=exp#jwt.decode

![image](https://user-images.githubusercontent.com/165497/153874984-1847e4fe-efe3-41bd-ab73-f6160097d4c5.png)

(in your case you would make them required by including them in `required_spec_claims`)

**Comment by Keats** [comment]
Looks like I was wrong! Can one of you do a PR?

**Comment by cloutiertyler** [comment]
@Keats Thanks very much! I'm happy to do a PR if @brocaar doesn't beat me to it!

**Comment by cloutiertyler** [comment]
Okay, I've submitted #240, however there's some details you should be aware of before accepting the PR. Maybe there's a better way we can do this.

I've now made `exp`, `nbf`, `iss`, `aud`, and, `sub` optional in accordance with the RFC, but because the library was previously requiring each of these if they were specified on the validation (e.g. `set_issuer(&["Keats"])`) even if they were not part of the `required_spec_claims`, if a developer is not aware of this change they may end up validating a missing `iss` because they thought they were requiring it.

For example, the following will now succeed, where it would previously fail.

```rust
let claims = json!({});
let mut validation = Validation::new(Algorithm::HS256);
validation.validate_exp = false;
validation.set_issuer(&["Keats"]);
let res = validate(deserialize_claims(&claims), &validation);
```

To get the same behavior as before you would have to do this:
```rust
let claims = json!({});
let mut validation = Validation::new(Algorithm::HS256);
validation.set_required_spec_claims(&["iss"]);
validation.validate_exp = false;
validation.set_issuer(&["Keats"]);
let res = validate(deserialize_claims(&claims), &validation);
```

It might be better instead to assume that all `spec_claim`s are required, unless they are specifically noted as optional or "not required". So for example you could do something like this:

```rust
let claims = json!({});
let mut validation = Validation::new(Algorithm::HS256);
validation.validate_exp = false;
validation.set_issuer(&["Keats"]);
let res = validate(deserialize_claims(&claims), &validation); // This will FAIL, just like before

let claims = json!({});
let mut validation = Validation::new(Algorithm::HS256);
validation.set_optional_spec_claims(&["iss"]);
validation.validate_exp = false;
validation.set_issuer(&["Keats"]);
let res = validate(deserialize_claims(&claims), &validation); // This will SUCCEED

let claims = json!({});
let mut validation = Validation::new(Algorithm::HS256);
validation.validate_exp = false;
let res = validate(deserialize_claims(&claims), &validation); // This will SUCCEED
```

Personally, I like the `required_spec_claims` better from an API design standpoint, but I just wanted to point out that there could be a security concern for people not paying close attention. Would be curious to hear your thoughts, @Keats @brocaar.

**Comment by brocaar** [comment]
@cloutiertyler @Keats: I think I prefer `required_spec_claims`. If that by default contains `exp`, then I think there is no impact when users update to a future version, as `exp` remains mandatory. From an API point of view, that might be better.

@Keats do you consider #240 mergable after the formatting and clippy issues have been fixed?

**Comment by Keats** [comment]
I think it's ok, have you tested the branch locally to make sure it works as you expect?

**Comment by brocaar** [comment]
@Keats I have created a new PR with the changes from @cloutiertyler  + the fmt and clippy fixes. I have also confirmed that this solves my issue.
